### ISSUE TITLE
(FLAN-T5) Scaling Instruction-Finetuned Language Models

### ISSUE THREAD

**Issue by eagle705** [eagle705/presentation#22]
# Note

# Author
- Hyung Won Chung 저술 (2022/12)
  - Google

# Abstract
- Finetuning language models on a collection of datasets phrased as instructions has been shown to improve model performance and generalization to unseen tasks
- explore instruction finetuning with a particular focus on 
  - (1) scaling the number of tasks, 
  - (2) scaling the model size, and 
  - (3) finetuning on chain-of-thought data. (CoT 데이터 좋다!)
- We find that instruction finetuning with the above aspects dramatically improves performance on 
  - a variety of model classes (PaLM, T5, U-PaLM)
  - prompting setups (zero-shot, few-shot, CoT),
  - evaluation benchmarks (MMLU, BBH, TyDiQA, MGSM, open-ended generation, RealToxicityPrompts).
- Flan-PaLM 540B instruction-finetuned on 1.8K tasks outperforms PaLM 540B by a large margin (+9.4% on average)
  - 1.8K tasks라니 엄청 많네
- We also publicly release Flan-T5 checkpoints,1 which achieve strong few-shot performance even compared to much larger models, such as PaLM 62B.

![image](https://user-images.githubusercontent.com/7252598/216892265-d3bde6cd-f61d-4301-8683-cdabd72effe9.png)

# Introduction
- In this paper we advance instruction finetuning in several ways.
  - First, we study the impact of **scaling** on instruction finetuning.
  - Second, we study the **effect of finetuning** on the ability of the models to perform **reasoning tasks.**
    - Our experiments show that whereas prior instruction finetuning methods that do not include chain-of-thought (CoT; Wei et al., 2022b) severely degrade performance on CoT evaluations, adding just nine CoT datasets into the finetuning mixture enables better performance on all evaluations. (9개의 CoT를 추가하면 더 좋아지고, CoT 없이 하면 리즈닝의 성능이 하락함)
- Based on these findings, we train Flan-PaLM by using a 540B-parameter model, increasing the number of finetuning tasks to 1.8K, and including CoT data. Flan-PaLM outperforms PaLM, achieving new state-of-the-art on several benchmarks
- In addition, we also instruction-finetune Flan-T5 models (80M to 11B)
- Flan-T5 11B outperforms T5 11B by double-digit improvements and even outperforms PaLM 62B on some challenging BIG-Bench tasks 
- 아래에 SC: self-consistency가 뭘까?
  - Xuezhi Wang, Jason Wei, Dale Schuurmans, Quoc Le, Ed Chi, and Denny Zhou. Self-consistency improves chain of thought reasoning in language models. arXiv preprint arXiv:2203.11171, 2022b. URL https://arxiv.org/abs/2203.11171
![image](https://user-images.githubusercontent.com/7252598/216893393-7ebb314e-771d-4753-b909-189a90af59f1.png)

# Flan Finetuning
- We instruction-finetune on a collection of data sources (Figure 2) with a variety of instruction template types (Figure 3). 
- We call this finetuning procedure Flan (Finetuning language models; Wei et al., 2021) and prepend “Flan” to the resulting finetuned models (e.g., Flan-PaLM)
![image](https://user-images.githubusercontent.com/7252598/216894201-1ec930d6-25c8-4d94-8715-59ac6de51e6a.png)

## Finetuning Data
- **Task mixtures**
  - In this paper we scale to 1,836 finetuning tasks by combining four mixtures from prior work: Muffin, T0-SF, NIV2, and CoT, as summarized in Figure 2
    - **Muffin** (80 tasks) comprises 62 tasks from Wei et al. (2021) and 26 new tasks that we added in this work, including dialog data
    - **T0-SF** (193 tasks) comprises tasks from T0 (Sanh et al., 2021) that do not overlap with the data used in Muffin (SF stands for “**sans Flan**”)
    - **NIV2** (1554 tasks) comprises tasks from Wang et al. (2022c)
- **Chain-of-thought finetuning mixture**
  - We create a new mixture of nine datasets from prior work for which **human** raters **manually wrote CoT** annotations for a training corpus
  - These nine datasets include tasks such as arithmetic reasoning (Cobbe et al., 2021), multi-hop reasoning (Geva et al., 2021), and natural language inference (Camburu et al., 2020). We manually compose ten instruction templates per task. A data card is in Appendix F.
- **Templates and formatting**
  - For Muffin, T0-SF, and NIV2, we use instructional templates for each task as given by the creators of the mixtures
  - For CoT, we manually write around ten instruction templates for each of the nine datasets.
    - 9개 데이터셋당 한개당 10개의 템플릿
  - To create few-shot templates, we write a variety of exemplar delimiters (e.g., "Q:"/"A:") and apply them randomly at the example level

![image](https://user-images.githubusercontent.com/7252598/216894703-f3f98689-8d4e-40dd-a7a8-96a95781b94e.png)

## Finetuning procedure
- We use a **constant learning rate schedule** and finetune using the **Adafactor** optimizer (Shazeer and Stern, 2018)
- We use packing (Raffel et al., 2020) to combine multiple training examples into a single sequence, separating inputs from targets using an end-of-sequence token.
  - packing을 한다는 표현이 자주 나오는데, 여러 examples을 하나의 single sequence로 넣는게 뭘까?
    - Masking is applied to prevent the tokens from attending to others across the packed example boundary.
    - 하나로 몰아서 넣고 masking으로 학습처리하는건가? packing하는건 GPU 활용때문인가
- The number of finetuning steps, learning rate, batch size, and dropout for each model are given in Appendix E.
- **Notably**, the amount of compute used for finetuning is only a small fraction relative to the training compute, as shown in Table 2. For example, we **only use 0.2% of the pre-training compute** to instruction-finetune Flan-PaLM 540B (approximately 512 v4 TPU chips for 37 hours)
  - 적은자원으로 finetuning을 한것!
  - We use the JAX-based T5X framework
    - JAX 썼네
![image](https://user-images.githubusercontent.com/7252598/216897009-bba1a327-7a9f-4f62-baf4-df91fa3517ec.png)

## Evaluation protocol
- **Evaluation benchmarks**
  - challenge benchmarks
    - (1) MMLU (Hendrycks et al., 2020) includes exam questions from 57 tasks such as mathematics, history, law, and medicine. 
    - (2) BBH includes 23 challenging tasks from BIG-Bench (Srivastava et al., 2022) for which PaLM performs below an average human rater (Suzgun et al., 2022). 
    - (3) TyDiQA (Clark et al., 2020) is a question-answering benchmark across 8 typologically diverse languages. 
    - (4) MGSM (Shi et al., 2022) is a multilingual benchmark of math word problems from Cobbe et al. (2021) manually translated into 10 languages.
- **Evaluation methods and metrics**
  - 생략

# Scaling to 540B parameters and 1.8K tasks
- We first examine the effect of scaling in terms of **(1) the size of model** and **(2) the number of finetuning tasks** on performance on held-out tasks
  - scale the model size by performing experiments on three PaLM model sizes: **8B, 62B, and 540B**
  - **sequentially add** task mixtures starting from the mixture with the fewest tasks to the mixture with the most tasks: **CoT, Muffin, T0-SF, and NIV2.**

![image](https://user-images.githubusercontent.com/7252598/216928182-27473fc8-963c-4c60-8d1f-b853999e1b50.png)

- First, we see that for all three model sizes shown, **multi-task instruction** finetuning improves performance by a large margin compared to no finetuning. The performance gain ranges from 9.4% to 15.5%.
- Second, **increasing the number of finetuning tasks** improves performance, although the majority of the improvement comes from using up to 282 tasks
- Finally, we see that increasing **model scale** by an order of magnitude (i.e., 8B → 62B or 62B → 540B) improves performance substantially for both finetuned and non-finetuned models

![image](https://user-images.githubusercontent.com/7252598/216928800-b278f8ec-0731-45d2-9c13-cab557b4fbfe.png)

# Finetuning with chain-of-thought annotations
- In this section we explore the effect of including chain-of-thought (CoT) data in the instruction finetuning mixture
- First, we show that the improved reasoning abilities of Flan-PaLM surprass prior models on several benchmarks.
  - Then, we ablate CoT finetuning data and show that whereas instruction finetuning without CoT actually degrades reasoning ability, including just nine CoT datasets improves performance on all evaluations.
- Finally, we show that CoT finetuning unlocks zero-shot reasoning via “let’s think step-by-step” on challenging BIG-Bench tasks.

## Finetuning on chain-of-thought improves reasoning on held-out tasks
- We first show that including nine datasets with chain-of-thought (CoT) annotations in the finetuning mixture improves reasoning ability. 
  - Table 4 also shows how CoT prompting can be combined with self-consistency
    - `Self-Consistency Improves Chain of Thought Reasoning in Language Models`
      - ![image](https://user-images.githubusercontent.com/7252598/217120546-43f41d35-fac7-4dfa-9242-1a015de0bc0c.png)

![image](https://user-images.githubusercontent.com/7252598/217120128-22c62ae8-a6c4-4e1b-8c8a-4bf6bbee74f0.png)

## Some chain-of-thought data is needed to maintain reasoning ability
- We next ablate the effect of including just nine CoT datasets in instruction finetuning
- We stratify evaluations into held-out CoT benchmarks (MMLU, BBH, and MGSM) and held-out non-CoT benchmarks (MMLU, BBH, and TyDiQA) and compute normalized averages for CoT and non-CoT
- 반반 섞은게 제일 성능이 높네..?! 신기하네
  - ![image](https://user-images.githubusercontent.com/7252598/217121429-7207aa24-4704-4d52-a5a3-47cda948fcf1.png)

## Unlocking zero-shot reasoning
- A final benefit of instruction finetuning on CoT data both with and without exemplars is that the resulting model is able to perform CoT reasoning in a zero-shot setting.
- Figure 6 shows that for the BBH benchmark of 23 unseen challenging BIG-Bench tasks, Flan-PaLM models can achieve improved performance by leveraging CoT reasoning activated by the phrase “let’s think step-by-step”
![image](https://user-images.githubusercontent.com/7252598/217121877-27015710-255f-49f6-a03e-25379d076c4a.png)

![image](https://user-images.githubusercontent.com/7252598/217122014-431ed29e-ca37-418b-9475-8576391c6317.png)

# Putting it all together
![image](https://user-images.githubusercontent.com/7252598/217122143-12360c46-30c6-4a55-80de-a9ee0919a670.png)

# Usability evaluation of open-ended generation
- (1) continuing to generate related text instead of answering a question, 
- (2) repeating the input question with minor modifications, and 
- (3) not knowing when to stop generating text. This is likely an artifact of not using end-of-sequence tokens in pre-training, and some examples of these errors are shown in Figure 9.

![image](https://user-images.githubusercontent.com/7252598/217122632-4dc286db-d064-42e2-b6a8-7496d7f655da.png)

![image](https://user-images.githubusercontent.com/7252598/217122838-dfc70998-1559-4a5e-818c-d36f5035289e.png)

# Discussion
In this work we extended instruction finetuning by (1) scaling the number of finetuning tasks, (2) scaling the size of the model, and (3) finetuning on CoT data.

- Scaling curves for instruction finetuning. 
- CoT finetuning is critical for reasoning abilities.
- Instruction finetuning generalizes across models.
- Instruction finetuning improves usability and mitigates some potential harms.
- Instruction finetuning is relatively compute-efficient. 

# Conclusions
-  (1) scaling to a 540B- parameter language model, 
- (2) scaling to 1.8K finetuning tasks, and 
- (3) including chain-of-thought (CoT) data in finetuning.
- Flan-PaLM achieves state-of-the-art performance on several benchmarks, such as 75.2% on five-shot MMLU. Flan-PaLM also has improved usability

# Appendix
![image](https://user-images.githubusercontent.com/7252598/217123168-630fb1b9-81ba-400b-8854-3f75fe0bab65.png)
